cuda/solver/gmres_kernels.cu

```diff
@@ -48,63 +57,687 @@ namespace cuda {
+    matrix::Dense<ValueType> *givens_sin, matrix::Dense<ValueType> *givens_cos,
+    Array<stopping_status> *stop_status, const size_type krylov_dim)
 {
-    NOT_IMPLEMENTED;
+    const auto num_threads = std::max(b->get_size()[0] * b->get_stride(),
+                                      krylov_dim * b->get_size()[1]);
+    const dim3 grid_dim(ceildiv(num_threads, default_block_size), 1, 1);
+    const dim3 block_dim(default_block_size, 1, 1);
+    constexpr auto block_size = default_block_size;
+
+    b->compute_norm2(b_norm);
+    initialize_1_kernel<block_size><<<grid_dim, block_dim>>>(
+        b->get_size()[0], b->get_size()[1], krylov_dim,
+        as_cuda_type(b->get_const_values()), b->get_stride(),
+        as_cuda_type(residual->get_values()), residual->get_stride(),
+        as_cuda_type(givens_sin->get_values()), givens_sin->get_stride(),
+        as_cuda_type(givens_cos->get_values()), givens_cos->get_stride(),
+        as_cuda_type(stop_status->get_data()));
 }
 
 GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_GMRES_INITIALIZE_1_KERNEL);
 
 
+// Must be called with at least `num_rows * stride_krylov` threads in total.
+template <size_type block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void initialize_2_1_kernel(
+    size_type num_rows, size_type num_rhs, size_type krylov_dim,
+    ValueType *__restrict__ krylov_bases, size_type stride_krylov,
+    ValueType *__restrict__ residual_norm_collection,
+    size_type stride_residual_nc)
+{
+    const auto global_id = blockIdx.x * blockDim.x + threadIdx.x;
+    const auto row_idx = global_id / stride_krylov;
+    const auto col_idx = global_id % stride_krylov;
+
+    if (row_idx < num_rows && col_idx < (krylov_dim + 1) * num_rhs) {
+        krylov_bases[row_idx * stride_krylov + col_idx] = zero<ValueType>();
+    }
+
+    if (row_idx < krylov_dim + 1 && col_idx < num_rhs) {
+        residual_norm_collection[row_idx * stride_residual_nc + col_idx] =
+            zero<ValueType>();
+    }
+}
+
+
+// Must be called with at least `num_rows * num_rhs` threads in total.
+template <size_type block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void initialize_2_2_kernel(
+    size_type num_rows, size_type num_rhs,
+    const ValueType *__restrict__ residual, size_type stride_residual,
+    const ValueType *__restrict__ residual_norm,
+    ValueType *__restrict__ residual_norm_collection,
+    ValueType *__restrict__ krylov_bases, size_type stride_krylov,
+    size_type *__restrict__ final_iter_nums)
+{
+    const auto global_id = blockIdx.x * blockDim.x + threadIdx.x;
+    const auto row_idx = global_id / num_rhs;
+    const auto col_idx = global_id % num_rhs;
+
+    if (global_id < num_rhs) {
+        residual_norm_collection[global_id] = residual_norm[global_id];
+        final_iter_nums[global_id] = 0;
+    }
+
+    if (row_idx < num_rows &&
+        col_idx < num_rhs) {  // global_id < num_rows * num_cols) {
```

Comment:
Is this commented code desired? I think the current version is clear enough.

Reply:
Yeah, saw that as well, forgot to remove the comment (it was commented out instead of deleted).